channels/chan_pjsip.c

```diff
@@ -727,6 +747,91 @@ static int answer(void *data)
+			? &channel->session->endpoint->media.codec_prefs_outgoing_offer
+			: &channel->session->endpoint->media.codec_prefs_outgoing_answer;
+		resolved_top = ast_stream_topology_create_resolved(
+			topology,
+			channel->session->pending_media_state->topology
+				? channel->session->pending_media_state->topology
+				: channel->session->endpoint->media.topology,
+			prefs,
+			&err_msg);
+		if (!resolved_top) {
+			SCOPE_EXIT_LOG_RTN_VALUE(-1, LOG_ERROR, "%s: Couldn't create resolved topology\n", ast_channel_name(ast));
+		}
+
+		for (index = 0; index < ast_stream_topology_get_count(resolved_top); ++index) {
+			if (ast_format_cap_empty(ast_stream_get_formats(ast_stream_topology_get_stream(resolved_top, index)))) {
+				if (prefs->transcode == CODEC_NEGOTIATION_TRANSCODE_ALLOW) {
+					ast_log(LOG_DEBUG, "%s No common codecs between channels and transcoding allowed. Still answering call.\n",
+						ast_sip_session_get_name(channel->session));
+					goto end;
+				} else {
+					const int response_code = 488;
+					struct indicate_data *ind_data = indicate_data_alloc(channel->session, AST_CONTROL_HANGUP, response_code, NULL, 0);
+					if (!ind_data) {
+						SCOPE_EXIT_LOG_RTN_VALUE(-1, LOG_ERROR, "%s: Couldn't alloc indicate data\n", ast_channel_name(ast));
+					}
+
+					if (ast_sip_push_task(channel->session->serializer, indicate, ind_data)) {
+						ast_log(LOG_ERROR, "%s: Cannot send response code %d to endpoint %s. Could not queue task properly\n",
+								ast_channel_name(ast), response_code, ast_sorcery_object_get_id(channel->session->endpoint));
+						ao2_cleanup(ind_data);
+					}
+					SCOPE_EXIT_LOG_RTN_VALUE(-1, LOG_WARNING, "%s: No common codecs between channels and transcoding not allowed. Sending 488 response.\n",
+						ast_sip_session_get_name(channel->session));
+				}
+			}
+		}
+
+		if (resolved_top && channel->session->pending_media_state->topology) {
+			struct regen_sdp_data *regen_data;
+			int is_offer;
+
+			ast_channel_set_stream_topology(ast, resolved_top);
+			ast_stream_topology_free(channel->session->pending_media_state->topology);
+			channel->session->pending_media_state->topology = ast_stream_topology_clone(resolved_top);
+
+			is_offer = pjmedia_sdp_neg_get_state(channel->session->inv_session->neg) == PJMEDIA_SDP_NEG_STATE_LOCAL_OFFER;
+			regen_data = regen_sdp_data_alloc(channel->session, is_offer);
+			if (!regen_data) {
+				ast_log(LOG_WARNING, "Failed to create resolved topology. Still answering call.\n");
+				goto end;
+			}
+			if (ast_sip_push_task(channel->session->serializer, regenerate_sdp, regen_data)) {
+				ast_log(LOG_WARNING, "Error regenerating SDP %s on channel %s\n", regen_data->is_offer ? "offer" : "answer", ast_channel_name(ast));
+				ao2_cleanup(regen_data);
+				goto end;
+			}
+		} else {
+			ast_log(LOG_NOTICE, "No resolved topology or pending_media_state topology. Still answering call.\n");
```

Comment:
``resolved_top`` is created but never freed in this case

Reply:
Thank you! I have fixed the leak and refactored the logic to make it simpler. Please mark as resolved if it looks good to you.